src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs

```diff
@@ -422,9 +450,17 @@ private void AddLexedTokenSlot()
             }
             else
             {
-                var tmp = new ArrayElement<SyntaxToken>[_lexedTokens.Length * 2];
-                Array.Copy(_lexedTokens, tmp, _lexedTokens.Length);
-                _lexedTokens = tmp;
+                var lexedTokens = _lexedTokens;
+
+                Array.Resize(ref _lexedTokens, _lexedTokens.Length * 2);
+
+                // Put lexedTokens back into the pool if it's correctly sized.
+                if (lexedTokens.Length == CachedTokenArraySize)
+                {
+                    // Fully clear out lexedTokens before releasing back to pool.
+                    Array.Clear(lexedTokens, 0, lexedTokens.Length);
```

Comment:
In `Dispose` we clear up to `_maxWrittenLexedTokenIndex`, but here we clear the entire length. Why the difference?

Reply:
There's no real reason other than I believe if we hit this codepath, then the two values are probably the same. But, it's clearer if we're consistent between the two places doing this, so I'll change this to match.

Reply:
Thanks. Consider factoring/extracting this common code

Reply:
Very reasonable request now that they are the same.